In [78]:
import pandas as pd

url= "https://breathecode.herokuapp.com/asset/internal-link?id=930&path=diabetes.csv"
total_data = pd.read_csv(url)
total_data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [79]:
total_data = total_data.drop_duplicates().reset_index(drop = True)
#elimina filas duplicadas y crea un nuevo indice secuencial, eliminando el anterior
total_data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [80]:
from sklearn.metrics import classification_report #accuracy, sensibility, specificity
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
import numpy as np

In [81]:
y = total_data['Outcome']
X = total_data.drop(columns='Outcome')
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.02, random_state=42)
total_data = total_data.drop_duplicates().reset_index(drop = True)
total_data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [82]:
from sklearn.feature_selection import SelectKBest #seleccion de mejores caracteristicas
from sklearn.model_selection import train_test_split #division de datos(train y test)

X = total_data.drop("Outcome", axis = 1) #creando un df igual pero sin '0utcome'
y = total_data["Outcome"] #variable que solo contendrá 'Outcome'

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)
#variable de flujo de trabajo que divide los datos, para que la prueba sea el 20% y asegura la replicabilidad

selection_model = SelectKBest(k = 7) #buscando las 7 caracteristicas mas relevantes
selection_model.fit(X_train, y_train) #hacemos el aprendizaje

selected_columns = X_train.columns[selection_model.get_support()] #nombres de las mejores 7 columnas

X_train_sel = pd.DataFrame(selection_model.transform(X_train), columns = selected_columns)
#tranformará X_train para contener selected_columns y lo guardamos en X_train_sel
X_test_sel = pd.DataFrame(selection_model.transform(X_test), columns = selected_columns)
#lo mismo pero en X_test

X_train_sel.head()

Pregnancies  Glucose  BloodPressure  Insulin   BMI  \
0          2.0     84.0            0.0      0.0   0.0   
1          9.0    112.0           82.0      0.0  28.2   
2          1.0    139.0           46.0     83.0  28.7   
3          0.0    161.0           50.0      0.0  21.9   
4          6.0    134.0           80.0    370.0  46.2   

   DiabetesPedigreeFunction   Age  
0                     0.304  21.0  
1                     1.282  50.0  
2                     0.654  22.0  
3                     0.254  65.0  
4                     0.238  46.0

In [83]:
X_test_sel.head()

Pregnancies  Glucose  BloodPressure  Insulin   BMI  \
0          6.0     98.0           58.0    190.0  34.0   
1          2.0    112.0           75.0      0.0  35.7   
2          2.0    108.0           64.0      0.0  30.8   
3          8.0    107.0           80.0      0.0  24.6   
4          7.0    136.0           90.0      0.0  29.9   

   DiabetesPedigreeFunction   Age  
0                     0.430  43.0  
1                     0.148  21.0  
2                     0.158  21.0  
3                     0.856  34.0  
4                     0.210  50.0

In [84]:
import os #importo biblioteca estandard para incluir la carpeta de abajo

X_train_sel["Outcome"] = y_train.values
#agrega nuevamente 'Outcome' al df(X_train_sel) con los valores de la caracteristicas y los resultados
X_test_sel["Outcome"] = y_test.values
#lo mismo pero en la prueba

# Crear el directorio ya que no existe
os.makedirs("../data/processed", exist_ok=True)

X_train_sel.to_csv("../data/processed/clean_train.csv", index = False)
X_test_sel.to_csv("../data/processed/clean_test.csv", index = False)

In [85]:
train_data = pd.read_csv("../data/processed/clean_train.csv")
test_data = pd.read_csv("../data/processed/clean_test.csv")

train_data.head()

Pregnancies  Glucose  BloodPressure  Insulin   BMI  \
0          2.0     84.0            0.0      0.0   0.0   
1          9.0    112.0           82.0      0.0  28.2   
2          1.0    139.0           46.0     83.0  28.7   
3          0.0    161.0           50.0      0.0  21.9   
4          6.0    134.0           80.0    370.0  46.2   

   DiabetesPedigreeFunction   Age  Outcome  
0                     0.304  21.0        0  
1                     1.282  50.0        1  
2                     0.654  22.0        0  
3                     0.254  65.0        0  
4                     0.238  46.0        1

In [86]:
X_train = train_data.drop(["Outcome"], axis = 1)
y_train = train_data["Outcome"]
X_test = test_data.drop(["Outcome"], axis = 1)
y_test = test_data["Outcome"]

In [87]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators = 60, random_state = 42)
model.fit(X_train, y_train)

RandomForestClassifier(n_estimators=60, random_state=42)

In [88]:
y_pred = model.predict(X_test)
y_pred

array([0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0,
       0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1,
       0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0])

In [89]:
from sklearn.metrics import accuracy_score

accuracy_score(y_test, y_pred)

0.7727272727272727

## BOOSTING:

In [90]:
from xgboost import XGBClassifier

model = XGBClassifier(n_estimators = 50, learning_rate = 0.1, max_depth = 3, random_state = 42)
model.fit(X_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              feature_weights=None, gamma=None, grow_policy=None,
              importance_type=None, interaction_constraints=None,
              learning_rate=0.1, max_bin=None, max_cat_threshold=None,
              max_cat_to_onehot=None, max_delta_step=None, max_depth=3,
              max_leaves=None, min_child_weight=None, missing=nan,
              monotone_constraints=None, multi_strategy=None, n_estimators=50,
              n_jobs=None, num_parallel_tree=None, ...)

In [91]:
y_pred = model.predict(X_test)
y_pred

array([1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1,
       0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1,
       0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0])

In [92]:
from sklearn.metrics import accuracy_score

accuracy_score(y_test, y_pred)

0.7402597402597403

### CONCLUSION:
- el modelo original de Random Forest tuvo un mejor rendimiento de 77% con respecto al boosting de un 74%.